### DISCUSSION TITLE
Learning Graphics Programming

### DISCUSSION THREAD

**Discussion by CaesarXInsanium** [sotrh/learn-wgpu#406] [Announcements]
I am sorry if this is not the correct place to ask this question
But I was wondering if anybody would recommend the best resources
to learn graphics programming, particularly with Rust.
The tutorial is amazing but it only covers how to use WGPU and it assumes knowledge of graphics programming concepts.

**Comment by CatCode79** [comment]
https://learnopengl.com/
^ despite being focused on opengl describes concepts of graphics programming valid to all APIs

https://fgiesen.wordpress.com/2011/07/09/a-trip-through-the-graphics-pipeline-2011-index/
^ despite the date the contents are still valid

You can learn further through the examples of the wgpu repository.

Unfortunately I do not remember rust resources on the subject. I read and learned something from the examples of the vulkano project, they are very descriptive but explain low-level concepts and strictly related to vulkan.

**Comment by sotrh** [comment]
I like https://learnopengl.com/, I also used https://www.opengl-tutorial.org/. Those both use C++, but they explain the topics pretty well.

Just out of curiosity, what graphics programming topics are you not familiar with?

**Comment by atgctg** [comment]
Would you also recommend https://webglfundamentals.org and https://webgl2fundamentals.org? Or are the contents of those not really related to WebGPU?

**Comment by sotrh** [comment]
The concepts are the same, but the code will be very different. Overall webgl is easier to start with. I'd recommend WebGL 2 if that's what you're interested in.

**Comment by CatCode79** [comment]
I found this interesting article:
[Primitive Geometry in wgpu and Rust](https://dev.to/whoisryosuke/primitive-geometry-in-wgpu-and-rust-1da1)
